### PR TITLE
ENT-12098: Added an initial implementation of an internal cfbs command to generate MPF release information

### PR DESCRIPTION
**Pull request by jakub-nt**

### CODE REVIEW

**Comment by cf-bottom** [comment]
Thank you for submitting a PR! Maybe @craigcomstock can review this?

File: masterfiles/analyze.py
Comment (olehermanse) [inline masterfiles/analyze.py:2]: Please put these files inside the `cfbs` package is installed as one self contained package. (`./cfbs/masterfiles/analyze.py`, `./cfbs/masterfiles/download_all_versions.py`, etc.)

File: cfbs/utils.py
Comment (olehermanse) [inline cfbs/utils.py:298]: Nice utility functions. Not very critical in this case, but you could add some unit tests to `./tests/test_utils.py`

File: cfbs/commands.py
Comment (olehermanse) [inline cfbs/commands.py:68]: When masterfiles is inside cfbs, it'll look like this:

```suggestion
from cfbs.masterfiles.generate_release_information import generate_release_information
```

File: masterfiles/check_download_matches_git.py
Comment (olehermanse) [inline masterfiles/check_download_matches_git.py:1]: This looks like a good case for a docstring. Also it's not just about supply chain attack, I'd go with a bit more generic description:

```suggestion
"""Check that the downloadable files match the git files

This can be used to monitor / detect if something has been changed, accidentally or maliciously.
"""
```

**Review by olehermanse — COMMENTED** [review]
Looks good so far, some preliminary comments.

File: cfbs/utils.py
Comment (jakub-nt) [reply cfbs/utils.py:298]: Done in https://github.com/cfengine/cfbs/pull/208/commits/2768a80ff7341b1d1e7330180f1f5e66ac771605

File: cfbs/commands.py
Comment (jakub-nt) [reply cfbs/commands.py]: Done in https://github.com/cfengine/cfbs/pull/208/commits/3f0f1ed42b9971a87d0bf816fe3e9ed416fd7371

File: masterfiles/analyze.py
Comment (jakub-nt) [reply masterfiles/analyze.py]: Done in https://github.com/cfengine/cfbs/pull/208/commits/3f0f1ed42b9971a87d0bf816fe3e9ed416fd7371

File: cfbs/masterfiles/analyze.py
Comment (craigcomstock) [inline cfbs/masterfiles/analyze.py:7]:
```suggestion
# ignore a path if it contains a component (single file or directory) from this list
```

File: cfbs/masterfiles/analyze.py
Comment (craigcomstock) [inline cfbs/masterfiles/analyze.py:10]: What is the TODO? The stuff above? I usually put the `TODO` before the description of the todo.

File: cfbs/masterfiles/generate_vcf_git_checkout.py
Comment (craigcomstock) [inline cfbs/masterfiles/generate_vcf_git_checkout.py:32]: Why is the check for git exists after a use of git? Seems like it should be up around line 25.

File: cfbs/masterfiles/generate_vcf_git_checkout.py
Comment (craigcomstock) [inline cfbs/masterfiles/generate_vcf_git_checkout.py:71]: Seems you should check that autoconf and automake are installed first so that autogen.sh can work.

Comment (craigcomstock) [reply cfbs/masterfiles/generate_vcf_git_checkout.py:71]: Since you are using `check=True` with `run()` but not catching any exceptions I worry about hard to read stack traces in case of problems. Probably fine in the exceptional cases but something like autoconf missing should be checked I think as that seems likely to happen.

**Review by craigcomstock — CHANGES_REQUESTED** [review]
Cool. Noticed a few things with a cursory look. I didn't check the logic too carefully. I can make another pass if need be.

File: cfbs/masterfiles/analyze.py
Comment (jakub-nt) [reply cfbs/masterfiles/analyze.py:7]: This one is on purpose, "iff" means "if and only if".

File: cfbs/masterfiles/analyze.py
Comment (jakub-nt) [reply cfbs/masterfiles/analyze.py:10]: Yes, sure, I'll change this.

File: cfbs/masterfiles/generate_vcf_git_checkout.py
Comment (jakub-nt) [reply cfbs/masterfiles/generate_vcf_git_checkout.py:32]: Oops, that's an oversight, in an earlier version checking in this place was sufficient. Will fix.

File: cfbs/masterfiles/analyze.py
Comment (jakub-nt) [reply cfbs/masterfiles/analyze.py]: Reworded in https://github.com/cfengine/cfbs/pull/208/commits/b516b19c038d149e10df342e60b71cb777cc34b5

File: cfbs/masterfiles/analyze.py
Comment (jakub-nt) [reply cfbs/masterfiles/analyze.py]: Rewritten in https://github.com/cfengine/cfbs/pull/208/commits/b516b19c038d149e10df342e60b71cb777cc34b5

File: cfbs/masterfiles/generate_vcf_git_checkout.py
Comment (jakub-nt) [reply cfbs/masterfiles/generate_vcf_git_checkout.py]: Fixed in https://github.com/cfengine/cfbs/pull/208/commits/b516b19c038d149e10df342e60b71cb777cc34b5

File: cfbs/masterfiles/generate_vcf_git_checkout.py
Comment (jakub-nt) [reply cfbs/masterfiles/generate_vcf_git_checkout.py:80]: Added checking for required commands in https://github.com/cfengine/cfbs/pull/208/commits/b516b19c038d149e10df342e60b71cb777cc34b5.

As for the stack traces, I've checked a few cases and I think it's readable enough. For the git command calls, the failing command and the status code should be enough information, and for the autogen and make calls, currently the entire stdout and stderr are printed.

**Review by craigcomstock — APPROVED** [review]
Nice.

File: cfbs/masterfiles/generate_vcf_git_checkout.py
Comment (larsewi) [inline cfbs/masterfiles/generate_vcf_git_checkout.py:28]: These two can probably be the same function. If the passed argument is an interable, then it can recursively call itself for each element in the iterable.

File: cfbs/masterfiles/generate_vcf_git_checkout.py
Comment (larsewi) [inline cfbs/masterfiles/generate_vcf_git_checkout.py:35]: Can you not rebase instead?

File: cfbs/masterfiles/generate_vcf_git_checkout.py
Comment (larsewi) [inline cfbs/masterfiles/generate_vcf_git_checkout.py:39]: Maybe we should handle exceptions if the command fails? I'm not sure.

File: cfbs/masterfiles/generate_vcf_git_checkout.py
Comment (larsewi) [inline cfbs/masterfiles/generate_vcf_git_checkout.py:61]: Leftover debug print?

**Review by larsewi — CHANGES_REQUESTED** [review]
Great work :rocket: I see you still have some `# TODO`'s left to address. I have some smaller nitpicks for you. But in general:
- It's nice if you try to avoid too many nested `if`'s. They then to make it really hard to understand the code. But this is not always the case.
- When you look for the masterfiles tarball URL, maybe it's possible to search for it, instead of hard coding where it is in the JSON? Maybe there is a common denominator you can look for? I'm not sure.

I tried running your program, but it hung. Or maybe it just takes a very long time? I did not try to figure out why.

Looking forward to re-review!

File: cfbs/masterfiles/download_all_versions.py
Comment (jakub-nt) [reply cfbs/masterfiles/download_all_versions.py:7]: I'll just remove it.

File: cfbs/masterfiles/check_tarball_checksums.py
Comment (jakub-nt) [reply cfbs/masterfiles/check_tarball_checksums.py:10]: Yes, this is an operator and it's `pathlib`'s equivalent of `os.path.join`. I'll just move everything to `os.path`.

File: cfbs/masterfiles/generate_vcf_git_checkout.py
Comment (larsewi) [inline cfbs/masterfiles/generate_vcf_git_checkout.py:93]: I don't think the second `f` is necessary here, since we don't do `git clone --recursive <REPO>`. However, it should not hurt either. So feel free to dismiss.
```suggestion
            ["git", "clean", "-dfx"],
```

**Review by larsewi — APPROVED** [review]
Thank you, this looks good to me :rocket:

File: cfbs/utils.py
Comment (larsewi) [inline cfbs/utils.py:318]: Remember to `close()` the file afterwards. Or you can use the `with` keyword to do it automatically

**Review by larsewi — COMMENTED** [review]
Great improvements :rocket: I have some minor comments and some open questions that I hope @olehermanse or @craigcomstock can answer

File: cfbs/masterfiles/check_download_matches_git.py
Comment (craigcomstock) [reply cfbs/masterfiles/check_download_matches_git.py:42]: I think user_error() which calls sys.exit() makes sense here. Maybe the message should suggest what to do!

What action would we take if the files don't match git? Fix git? Fix the downloadable files? Fix something in this codebase that understands some quirk of a particular release?

File: cfbs/masterfiles/generate_vcf_git_checkout.py
Comment (craigcomstock) [reply cfbs/masterfiles/generate_vcf_git_checkout.py:55]: I might suggest that we FIX these tags that are exceptions here. It would seem of not much consequence since all of these versions are out of support.

File: cfbs/masterfiles/generate_vcf_git_checkout.py
Comment (craigcomstock) [reply cfbs/masterfiles/generate_vcf_git_checkout.py:90]: I do think that `make dist` makes a lot of sense to me.

File: cfbs/utils.py
Comment (olehermanse) [reply cfbs/utils.py]: +1 for using context managers (`with`)

**Review by larsewi — APPROVED** [review]
LGTM :rocket: @jakub-nt maybe you can squash the commits? @olehermanse do you want to have one last look before we hit the merge button?

File: cfbs/masterfiles/generate_vcf_git_checkout.py
Comment (olehermanse) [reply cfbs/masterfiles/generate_vcf_git_checkout.py:90]: It could be something to do in addition to, not instead of. We explicitly want to compare what's in git here.

**Review by olehermanse — APPROVED** [review]
👍 Looks good to me, merging, and we can always submit follow-up fixes in new PRs.

File: cfbs/masterfiles/check_download_matches_git.py
Comment (jakub-nt) [reply cfbs/masterfiles/check_download_matches_git.py:42]: https://github.com/cfengine/cfbs/pull/212 makes the change to no longer exit on the first differing version.

File: cfbs/masterfiles/download_all_versions.py
Comment (jakub-nt) [reply cfbs/masterfiles/download_all_versions.py:86]: https://github.com/cfengine/cfbs/pull/208#discussion_r1862058268

**Pull request merged**
